### PR TITLE
Default motor PWM configuration

### PR DESCRIPTION
**Pull request by MaEtUgR**
### Solved Problem
PWM configuration parameters for ESCs have ambiguous meanings depending on the use case and have to manually be adjusted according to the context that the allocator already has.

Fixes https://github.com/PX4/PX4-Autopilot/issues/21506

### Solution
- Default to PWM 1000us for disarmed (servos still have 1500us)
  PWM 900us is out of spec of any ESC I know and they all stop the motor with 1000us
- Default to a range of PWM 1100us - 1900us for motors
  This is the most compatible range I can think of with any ESC I worked with. Some have a lower limit of 1050us or 1075us but all turn the motor with 1100us. Some stop the range at 1925us or 1940us but all of them have the range not yet ended with 1900us. The idea is that the conservative defaults work reliably out of the box. If a user carefully follows the steps and e.g. experimentally finds out the exact PWM limits for all ESCs in use they can still configure that. From looking at logs I see that most users don't do that.
- Fixed wings deliberately turn off the forward motors if they are commanded to produce zero thrust.
- StandardVTOL do the same as fixed wings when not hovering
- Tailsitter can now stop motor in fast forward flight, bench and simulation tested
- [x] Tiltrotor simulation test
- Rover ackermann & differential simulation tested only
- [x] Boat simulation test
- [x] Submarine simulation test

### Changelog Entry
For release notes:
```
Attention: Default disarmed PWM was changed from 900us to 1000us!
Improvement: Default motor PWM limits minimum 1100 maximum 1900
Documentation: https://docs.px4.io/main/en/advanced_config/esc_calibration.html
```
Documentation improvement pending: https://github.com/PX4/PX4-user_guide/pull/2599

### Alternatives
I also thought about keeping the ambiguity of PWM configuration and adding some magic again to change it automatically but just for multicopters in one way and for fixed wings the other way. But I think that's not the right solution because it will cause more confusion and corner cases and explicit configuration and use.

### Test coverage
- I tested this on a quadcopter with PWM ESCs both configured as multirotor, fixed-wing and StandardVTOL with 1 or more pushers. It behaves as I expect in these cases.
- ~What's still missing are the VTOL cases. I think I can figure out standard VTOL but might need some help with Tiltrotor, let's see.~ **EDIT:** See above what I covered.

### Context
1.14 release has only the new actuator configuration now and I want to make sure we get these PWM problems out of the way from the beginning since many people will switch and reconfigure the actuators once.

### CODE REVIEW

**Comment by MaEtUgR** [comment]
StandardVTOL now works as expected. Other VTOL cases 👀
It looks like a tailsitter was not even capable to turn off the motor before judging by this comment: https://github.com/PX4/PX4-Autopilot/blob/0afd93a9e9696be0f86dd99c7e9fa9b04066df35/src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessTailsitterVTOL.cpp#L100

File: src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessFixedWing.cpp [hunk 1]
```diff
@@ -61,6 +62,25 @@ ActuatorEffectivenessFixedWing::getEffectivenessMatrix(Configuration &configurat
 	return (rotors_added_successfully && surfaces_added_successfully);
 }
 
+void ActuatorEffectivenessFixedWing::updateSetpoint(const matrix::Vector<float, NUM_AXES> &control_sp,
+		int matrix_index, ActuatorVector &actuator_sp, const matrix::Vector<float, NUM_ACTUATORS> &actuator_min,
+		const matrix::Vector<float, NUM_ACTUATORS> &actuator_max)
+{
+	// Stop front facing motors when they are commanded 0 thrust
```
Comment (sfuhrer) [inline src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessFixedWing.cpp:69]: What does this change to the current implementation? It now goes to the configured MIN_PWM when thrust is zero, default 1000: 
image

File: src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessStandardVTOL.cpp [hunk 2]
```diff
@@ -83,6 +84,27 @@ void ActuatorEffectivenessStandardVTOL::allocateAuxilaryControls(const float dt,
 	}
 }
 
+void ActuatorEffectivenessStandardVTOL::updateSetpoint(const matrix::Vector<float, NUM_AXES> &control_sp,
+		int matrix_index, ActuatorVector &actuator_sp, const matrix::Vector<float, NUM_ACTUATORS> &actuator_min,
+		const matrix::Vector<float, NUM_ACTUATORS> &actuator_max)
+{
+	if (matrix_index == 0) {
+		// Stop front facing motors when they are commanded 0 thrust
```
Comment (sfuhrer) [inline src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessStandardVTOL.cpp:92]: Don't we just want to keep them at PWM_MIN? Stop for me is to, well, stop completely, which is what most VTOL pusher/pullers should be confiugred to do, but if one wants to have them always spinning that should be possible by increasing PWM_MIN to higher than 1000 right?

File: src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessStandardVTOL.cpp [hunk 3]
```diff
@@ -94,13 +116,13 @@ void ActuatorEffectivenessStandardVTOL::setFlightPhase(const FlightPhase &flight
...
 	case FlightPhase::FORWARD_FLIGHT:
-		_stopped_motors = _mc_motors_mask;
+		_stopped_motors |= _mc_motors_mask;
 		break;
 
 	case FlightPhase::HOVER_FLIGHT:
 	case FlightPhase::TRANSITION_FF_TO_HF:
 	case FlightPhase::TRANSITION_HF_TO_FF:
-		_stopped_motors = 0;
+		_stopped_motors &= ~_mc_motors_mask;
```
Comment (sfuhrer) [inline src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessStandardVTOL.cpp:125]: Would this stop the pusher in hover? It's needed for pusher support.

**Review by sfuhrer — COMMENTED** [review]
Thanks for this push, I think it makes sense to align all this.

I didn't yet have time to in depth go through it in detail, but maybe answering my couple of questions already is enough for me to get a better understanding.

File: src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessFixedWing.cpp (hunk 1 above)
Comment (MaEtUgR) [reply src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessFixedWing.cpp:69]: It defaults to disarmed 1000, min 1100, max 1900 for **any** motor. And instead of setting the PWM_MIN back to 1000 for specific use cases the plan is to explicitly stop motors when desired with the existing infrastructure so a default fixed-wing pusher has PWM range 1100 - 1900 but can explicitly turn off the motor with low enough commanded thrust. So it's very clear that if you command a thrust the motor turns and if you command it to stop it stops and doesn't depend on the ESC e.g. PWM config, DroneCAN, DSHOT, ...

File: src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessStandardVTOL.cpp (hunk 2 above)
Comment (MaEtUgR) [reply src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessStandardVTOL.cpp:92]: > which is what most VTOL pusher/pullers should be confiugred to do

That's what I'm trying to solve that you have to configure PWM according to what the motor is supposed to do which is context the allocator has but you currently need to set manually.

> if one wants to have them always spinning that should be possible by increasing PWM_MIN to higher than 1000 right?

It's possible and currently, every user has to manually set the PWM_MIN of any motor that should not stop spinning e.g. any multicopter motor to a higher value. Also consider this only works for PWM outputs and the special cases all have to be manually configured for any other interface. E.g. DSHOT pusher is not even possible because the functionality is captured manually in PWM configuration. Instead of re-adding some magic to guess which motor should not stop spinning without having the context I suggest configuring every motor to be the same way and the allocator to have the interface available to command a motor to stop depending on what it does. So in summary it's extending the VTOL motor start/stop interface to all cases a motor should start/stop.

File: src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessStandardVTOL.cpp (hunk 3 above)
Comment (MaEtUgR) [reply src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessStandardVTOL.cpp:125]: > Would this stop the pusher in hover? It's needed for pusher support.

No, exactly. That's what I took care of. It only enables and disables mc motors but the pusher is only off when it's commanded zero thrust. This can be extended to where the controller decides to always run it even if just idling but currently for 1.14, it just works again like it was.

**Comment by mrpollo** [comment]
Thanks for carrying this through the finish line before the release @MaEtUgR 

https://media.giphy.com/media/Mliueouehmpag/giphy.gif

**Comment by MaEtUgR** [comment]
ok, I went through all vehicle types including boat, uuv, tiltrotor in simulation and made sure it does what I expect:
Stop motors with <2% output command if they are not used for hovering. I also went through to make naming consistent and reduce duplication. It's in a reviewable state now.

As discussed with @bkueng the next step would be to allow the controller to decide which motors are turning off or not but we need to discuss the interface there and I didn't want to introduce that bigger change for 1.14 now.

**Review by sfuhrer — COMMENTED** [review]
What I see makes sense to me. As discussed, let's though give this some testflying on all VTOL types early next week. That's something I anyway had planned to do to sign off the release.

**Comment by dagar** [comment]
> Default to PWM 1000us for disarmed (servos still have 1500us)
> PWM 900us is out of spec of any ESC I know and they all stop the motor with 1000us

I'm not sure, maybe we should skip this one until we fix PWM calibration.

**Comment by MaEtUgR** [comment]
Just rebased on main again since @sfuhrer and I go out to flight test the latest state of main including this.

**Comment by sfuhrer** [comment]
Flight tested on a MC, FW, VTOL tailsitter, VTOL standard, VTOL tiltrotor.
logs:
VTOL tailsitter: https://review.px4.io/plot_app?log=bf5e796d-1fe0-44be-b04e-6ca1456422c4
VTOL standard: https://review.px4.io/plot_app?log=dd707948-f84a-449d-b90c-e748497b1cf2
FW: https://review.px4.io/plot_app?log=2e7daeca-9281-410c-9288-a154602d42a4

No issues around the new PWM logic as proposed in this Pr was found. All VTOLs switched off the motors not in use correctly. On one vehicle (the tiltrotor), the ESCs required a minimum PWM of 1150 to turn, and the ESC was not configurable. On all other vehicles the default range (1100-1900 for motors) was kept. I didn't check if the ESC response would have gone above 1900.

**Review by sfuhrer — APPROVED** [review]

**Comment by MaEtUgR** [comment]
For this one I'm looking into making sure all the PWM calibrations are always done with 2000us and then 1000us as the reference values which right now they are the configured `PWM_MIN` and `PWM_MAX` values which mostly leads to unexpected calibration results.

**Comment by MaEtUgR** [comment]
Now that #21711 is merged PWM outputs can consistently be calibrated on 1.14. So I:
- rebased this pr on main branch cleanly
- fixed a comment typo here:
https://github.com/PX4/PX4-Autopilot/compare/3ca90ff8402ceca9dbe1015f9720e35a503272ff..bee2f4bcc4bfd06a4ee7918ab7aa42131b2cc130#diff-38572b66bed99f09ed91288986ccddfb67a5bac6b1eff49ec6851e9d342aa286R228
- Added a commit to remove unused PWM defines that still contained the old range which could be confusing see bee2f4b

**Comment by mrpollo** [comment]
Let's make sure the release notes are calling out any user actions for upgrading

**Comment by MaEtUgR** [comment]
Here's the documentation change for this mainly with the fixed PWM calibration (https://github.com/PX4/PX4-Autopilot/pull/21711): https://github.com/PX4/PX4-user_guide/pull/2599

That's what we need to reference in the release notes.

**Comment by MaEtUgR** [comment]
I think https://github.com/PX4/PX4-user_guide/pull/2599 should be ready now. So once the doc is available via URL we can merge this and notify + reference what should be done.

**Comment by vincentpoont2** [comment]
@MaEtUgR Does this apply to all Airframes?

**Comment by mengchaoheng** [comment]
@MaEtUgR In fixed-wing, why can I arm the vehicle even if the throttle stick is not at the lowest point?

**Comment by sfuhrer** [comment]
@mengchaoheng please do not use random PRs for issue reports - open a new issue and/or ask on discord or discuss. Thanks!
